quarkchain/cluster/slave.py

```diff
@@ -887,26 +887,23 @@ def __get_shard_size(self):
     def start(self):
         self.loop.create_task(self.__start_server())
 
-    def start_and_loop(self):
-        self.start()
+    def do_loop(self):
         try:
-            self.loop.run_forever()
+            self.loop.run_until_complete(self.shutdown_future)
         except KeyboardInterrupt:
             pass
-        self.shutdown()
 
     def shutdown(self):
-        if self.shutdown_in_progress:
-            return
+        if not self.shutdown_future.done():
+            self.shutdown_future.set_result(None)
 
-        self.shutdown_in_progress = True
-        if self.master is not None:
-            self.master.close()
+        # if self.master is not None:
```

Comment:
in my understanding, shutdown is usually caused by master connection error, which means `MasterConnection.close` will invoke `SlaveServer.shutdown`. amiright? safe to remove these two lines? (otherwise infinite recursion

Reply:
@qcgg